modules/core/src/main/scala/parser.scala

```diff
@@ -251,13 +253,43 @@ object GraphQLParser {
       sel  <- SelectionSet
     } yield Ast.FragmentDefinition(name, cond, dirs, sel)
 
-  lazy val FragmentName =
+  lazy val FragmentName: Parser[Ast.Name] =
     Name.filter(_ != Ast.Name("on"))
 
-  lazy val TypeCondition =
-    keyword("on") ~> NamedType
+  lazy val TypeCondition: Parser[Ast.Type.Named] =
+    keyword("on") *> NamedType
+
+  lazy val Value: Parser[Ast.Value] = recursive[Ast.Value] { rec =>
+
+    lazy val NullValue: Parser[Ast.Value.NullValue.type] =
+      keyword("null").as(Ast.Value.NullValue)
+
+    lazy val EnumValue: Parser[Ast.Value.EnumValue] =
+      Name.filter {
```

Comment:
I think it would be bettter to have `Name` not parse true, false or null. so you don't have a consuming failure here (which filter will create).

`not(string("true") | string("false") | string("null")) *> ...` could be how `Name` is defined.